dns/handler_whitelist_test.go

```diff
@@ -0,0 +1,187 @@
+	"net"
+	"testing"
+
+	"github.com/miekg/dns"
+	"github.com/mysteriumnetwork/node/core/policy"
+	"github.com/mysteriumnetwork/node/firewall"
+	"github.com/mysteriumnetwork/node/market"
+	"github.com/stretchr/testify/assert"
+)
+
+var (
+	policyDNSZone      = market.AccessPolicy{ID: "wildcard-domain"}
+	policyDNSZoneRules = market.AccessPolicyRuleSet{
+		ID: "wildcard-domain",
+		Allow: []market.AccessRule{
+			{Type: market.AccessPolicyTypeDNSZone, Value: "wildcard.com"},
+		},
+	}
+
+	policyDNSHostname      = market.AccessPolicy{ID: "domain"}
+	policyDNSHostnameRules = market.AccessPolicyRuleSet{
+		ID: "domain",
+		Allow: []market.AccessRule{
+			{Type: market.AccessPolicyTypeDNSHostname, Value: "single.com"},
+		},
+	}
+)
+
+func Test_WhitelistAnswers(t *testing.T) {
+	tests := []struct {
+		response       *dns.Msg
+		whitelistedIPs map[string]int
+		message        string
+	}{
+		{
+			&dns.Msg{
+				Answer: []dns.RR{
+					&dns.A{
+						Hdr: dns.RR_Header{Name: "single.com.", Rrtype: dns.TypeA, Class: dns.ClassINET, Ttl: 0},
+						A:   net.ParseIP("0.0.0.3"),
+					},
+					&dns.A{
+						Hdr: dns.RR_Header{Name: "single.com.", Rrtype: dns.TypeA, Class: dns.ClassINET, Ttl: 0},
+						A:   net.ParseIP("0.0.0.4"),
+					},
+				},
+			},
+			map[string]int{
+				"0.0.0.3": 1,
+				"0.0.0.4": 1,
+			},
+			"should allow whitelisted hostname",
+		},
+		{
+			&dns.Msg{
+				Answer: []dns.RR{
+					&dns.A{
+						Hdr: dns.RR_Header{Name: "cdn.single.com.", Rrtype: dns.TypeA, Class: dns.ClassINET, Ttl: 0},
+						A:   net.ParseIP("0.0.0.2"),
+					},
+				},
+			},
+			map[string]int{},
+			"should not allow zone of whitelisted hostname",
+		},
+		{
+			&dns.Msg{
+				Answer: []dns.RR{
+					&dns.A{
+						Hdr: dns.RR_Header{Name: "belekas.com.", Rrtype: dns.TypeA, Class: dns.ClassINET, Ttl: 0},
+						A:   net.ParseIP("0.0.0.1"),
+					},
+				},
+			},
+			map[string]int{},
+			"should not allow unknown hostname",
+		},
+
+		{
+			&dns.Msg{
+				Answer: []dns.RR{
+					&dns.A{
+						Hdr: dns.RR_Header{Name: "wildcard.com.", Rrtype: dns.TypeA, Class: dns.ClassINET, Ttl: 0},
+						A:   net.ParseIP("0.0.0.8"),
+					},
+					&dns.A{
+						Hdr: dns.RR_Header{Name: "wildcard.com.", Rrtype: dns.TypeA, Class: dns.ClassINET, Ttl: 0},
+						A:   net.ParseIP("0.0.0.9"),
+					},
+				},
+			},
+			map[string]int{
+				"0.0.0.8": 1,
+				"0.0.0.9": 1,
+			},
+			"should allow whitelisted wildcard hostname",
+		},
+		{
+			&dns.Msg{
+				Answer: []dns.RR{
+					&dns.A{
+						Hdr: dns.RR_Header{Name: "cdn.wildcard.com.", Rrtype: dns.TypeA, Class: dns.ClassINET, Ttl: 0},
+						A:   net.ParseIP("0.0.0.6"),
+					},
+					&dns.A{
+						Hdr: dns.RR_Header{Name: "cdn.wildcard.com.", Rrtype: dns.TypeA, Class: dns.ClassINET, Ttl: 0},
+						A:   net.ParseIP("0.0.0.7"),
+					},
+				},
+			},
+			map[string]int{
+				"0.0.0.6": 1,
+				"0.0.0.7": 1,
+			},
+			"should allow zone of whitelisted wildcard hostname",
+		},
+	}
+
+	for _, test := range tests {
```

Comment:
Do parameterized test cases appear properly named in test runners, e.g. IDE?
Usually, when doing data-driven tests, they come in the form of:

```
	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
```

Reply:
Added